tests/containers/artifacts.py

```diff
@@ -7,93 +7,131 @@
 
-    self.assertEqual(attribute_container.name, 'SystemRoot')
+    expected_attribute_names = ['case_sensitive', 'name', 'value']
 
-    expected_dict = {
-        'case_sensitive': False,
-        'name': 'SystemRoot',
-        'value': 'C:\\Windows'}
-
-    test_dict = attribute_container.CopyToDict()
-
-    self.assertEqual(test_dict, expected_dict)
+    attribute_names = sorted(attribute_container.GetAttributeNames())
+    self.assertEqual(attribute_names, expected_attribute_names)
 
 
 class HostnameArtifactTest(shared_test_lib.BaseTestCase):
   """Tests for the hostname artifact."""
 
-  # TODO: replace by GetAttributeNames test
-  def testCopyToDict(self):
-    """Tests the CopyToDict function."""
-    attribute_container = artifacts.HostnameArtifact(name='mydomain.com')
+  def testGetAttributeNames(self):
+    """Tests the GetAttributeNames function."""
+    attribute_container = artifacts.HostnameArtifact()
 
-    self.assertEqual(attribute_container.name, 'mydomain.com')
+    expected_attribute_names = ['name', 'schema']
 
-    expected_dict = {
-        'name': 'mydomain.com',
-        'schema': 'DNS'}
+    attribute_names = sorted(attribute_container.GetAttributeNames())
+    self.assertEqual(attribute_names, expected_attribute_names)
 
-    test_dict = attribute_container.CopyToDict()
 
-    self.assertEqual(test_dict, expected_dict)
+class OperatingSystemArtifactTest(shared_test_lib.BaseTestCase):
+  """Tests for the operating system artifact."""
 
+  def testNormalizedProduct(self):
+    """Tests the normalized_product property."""
+    attribute_container = artifacts.OperatingSystemArtifact(
+        product='Windows Server 2012 R2 Standard')
 
-class SystemConfigurationArtifactTest(shared_test_lib.BaseTestCase):
-  """Tests for the system configuration artifact."""
+    self.assertEqual(attribute_container.normalized_product, 'Windows 2012')
 
-  # TODO: replace by GetAttributeNames test
-  def testCopyToDict(self):
-    """Tests the CopyToDict function."""
-    attribute_container = artifacts.SystemConfigurationArtifact(
-        code_page='cp1252', time_zone='UTC')
+  def testVersionTuple(self):
+    """Tests the version_tuplele property."""
+    attribute_container = artifacts.OperatingSystemArtifact(version="5.1")
+    self.assertEqual(attribute_container.version_tuple, (5, 1))
 
-    self.assertEqual(attribute_container.time_zone, 'UTC')
+    attribute_container = artifacts.OperatingSystemArtifact()
+    self.assertIsNone(attribute_container.version_tuple)
 
-    expected_dict = {
-        'code_page': 'cp1252',
-        'time_zone': 'UTC',
-        'user_accounts': []}
+    attribute_container = artifacts.OperatingSystemArtifact(version="5.a")
+    self.assertIsNone(attribute_container.version_tuple)
 
-    test_dict = attribute_container.CopyToDict()
+  def testCompare(self):
+    """Tests the Compare function."""
+    attribute_container1 = artifacts.OperatingSystemArtifact(
```

Comment:
Rather than reusing "attribute_container1" and "attribute_container2" in multiple asserts, please rename this variables so these tests are easier to read - self.assertFalse(win2k12_artifact.Compare(winxp_artifact)) is a lot easier to reason about than self.assertFalse(attribute_container1.Compare(attribute_container2))

Reply:
Seeing this has come up multiple times, I opt to add this to the style guide: https://github.com/log2timeline/plaso/issues/2303

Reply:
This is already addressed in the google style guide: "Function names, variable names, and filenames should be descriptive". `attribute_container1` is less descriptive than `win2k12_artifact`. 

We could debate "sufficiently descriptive", but I don't think we could write a style guide definition without a lot of pain. We'll have to keep discussing this in code review, IMO.

Reply:
this remains subjective interpret-able, for me the type is more descriptive than the function, especially for tests. This needs to be clarified in more detail in the style guide.